c10/core/TensorImpl.cpp

```diff
@@ -93,20 +96,28 @@ void TensorImpl::release_resources() {
 }
 
 int64_t TensorImpl::dim() const {
+  TORCH_CHECK(type_id_ != UndefinedTensorId(), type_id_, " does not have dim");
   return sizes_.size();
 }
 
 int64_t TensorImpl::size(int64_t d) const {
+  TORCH_CHECK(type_id_ != UndefinedTensorId(), type_id_, " does not have size");
   d = at::maybe_wrap_dim(d, dim(), false);
   return sizes_[d];
 }
 
 int64_t TensorImpl::stride(int64_t d) const {
+  TORCH_CHECK(type_id_ != UndefinedTensorId() && layout() == kStrided, type_id_, " does not have stride");
   d = at::maybe_wrap_dim(d, dim(), false);
   return strides_[d];
 }
 
 TensorImpl* TensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
+  // We only allow this operation on dense tensors
+  if (!(has_storage() && layout() == kStrided)) {
```

Comment:
same as before, is there a difference between has_storage and the layout being strided?

Reply:
I guess my main question is -- can we just has_storage -> storage_ != NULL as an equivalence for kStrided?  I don't know if that works with C2 tensors or not, though.